zipline/pipeline/loaders/blaze/core.py

```diff
@@ -1011,12 +1011,47 @@ def last_in_date_group(df, reindex, have_sids=have_sids):
                     )
                 else:
                     last_in_group = last_in_group.reindex(dates)
+            # Unstack will fill all missing values with NaN; we need to fix
+            # this for all types that are not float.
+            if not df.empty:
+                for column in columns:
+                    if isinstance(column.missing_value, NotSpecifiedType):
+                        raise ValueError("Must specify a missing value for "
+                                         "column %d." % column.name)
+                    # Need special logic in case we're replacing with Nones
+                    elif df[column.name].dtype == categorical_dtype:
+                        last_in_group[column.name] = last_in_group[
+                            column.name
+                        ].where(pd.notnull(last_in_group[column.name]),
+                                column.missing_value)
+                    # Need to convert from float col to datetime col
+                    elif df[column.name].dtype == datetime64ns_dtype:
```

Comment:
The next two lines here are:

```
last_in_group[column.name] = last_in_group[
 +                            column.name
 +                        ].astype('datetime64[ns]')
 +                        last_in_group[column.name] = last_in_group[
 +                            column.name
 +                        ].fillna(column.missing_value)
```
We're pulling data out of the frame, changing its type, and writing it back into the frame.
We're then pulling the data we just wrote back **out** of the frame, ffilling it, and writing it in again.

So, two questions:

1. Why is the `astype` only necessary in the datetime64 case?
2. Can we avoid writing twice here (e.g. by doing `last_in_group[...] = `last_in_group[...].astype(...).fillna(...)`)?